[Distributed] Add custom allreduce support for ROCM

ilmarkov opened the pull request:
Enable custom allreduce for AMD gpus (MI300X)
Fix custom allreduce test
Set `RAY_RUNTIME_ENV_IGNORE_GITIGNORE` in tests utils that use ray as it ignores `.so` files uploading sources to ray workers.

Commit:
Add custom allreduce support for ROCM

Signed-off-by: ilmarkov <imarkov@redhat.com>

Commit:
Fix

Signed-off-by: ilmarkov <imarkov@redhat.com>

Commit:
Fix test

Signed-off-by: ilmarkov <imarkov@redhat.com>

Commit:
Fix distributed tests on AMD

Remove working_dir from ray init as it caused import errors

Signed-off-by: ilmarkov <imarkov@redhat.com>

Commit:
Remove old comment

Signed-off-by: ilmarkov <imarkov@redhat.com>

SageMoore commented on vllm/distributed/device_communicators/rocm_wrapper.py:
```diff
@@ -0,0 +1,175 @@
+
+def find_loaded_library(lib_name) -> Optional[str]:
+    """
+    According to according to https://man7.org/linux/man-pages/man5/proc_pid_maps.5.html,
```
Nit: Small typo "According to"

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -169,36 +174,84 @@ static DINLINE FlagType ld_flag_volatile(FlagType* flag_addr) {
+  // testing. Might be the case that hardware provides stronger guarantee than
+  // the memory model.
+  uint32_t flag = self_sg->_flag[blockIdx.x] + 1;
+#if !defined(USE_ROCM)
```
Do you think it makes more sense to move the ifdef's to outside of start_sync and end_sync? That way the two implementations are completely separate.

ilmarkov replied:
If you think it is going to improve code readability, sure

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -22,24 +27,22 @@
-  // may write counter+1 while current GPU is busy waiting for counter. We use
-  // alternating counter array to avoid this possibility.
-  alignas(128) FlagType peer_counter[2][kMaxBlocks][8];
+  alignas(128) FlagType start[kMaxBlocks][8];
```
I think a longer comment is warranted here. Especially since you are deleting one :).

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -22,24 +27,22 @@
 struct __align__(16) RankData {
-  const void* __restrict__ ptrs[8];
+  const void*
+#if !defined(USE_ROCM)
```
Comment?

ilmarkov replied:
I double checked it. The  `__restrict__` keyword within a struct does not have a performance implications. According to this [answer](https://forums.developer.nvidia.com/t/restrict-seems-to-be-ignored-for-base-pointers-in-structs-having-base-pointers-with-restrict-as-kernel-arguments-directly-works-as-expected/154020/13) there is no effect of it in `nvcc`. `hipcc` in its turn raises a compiler error when the keyword is used. So I will remove the keyword. for both platforms in this case.

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -220,13 +273,13 @@ __global__ void __launch_bounds__(512, 1)
     ((P*)result)[idx] = packed_reduce<P, ngpus, A>((const P**)&dp.ptrs[0], idx);
   }
-  multi_gpu_barrier<ngpus, false>(sg, self_sg, rank);
+  end_sync<ngpus, true>(sg, self_sg, rank);
```
Could you explain why you changed the boolean value here?

ilmarkov replied:
The template parameter meaning has changed. It is opposite now. "True" used to mean here that we want to do a memory fence, now it means that we don't have to do any additional thread sync as this is the end of the kernel.

tlrmchlsmth replied:
Small nit on naming: The names `start_sync` and `end_sync` makes me think that we're starting and ending the synchronization, and that both are needed to make a barrier -- but IIUC, 
it should be read as `sync_at_start` and `sync_at_end`, so consider renaming the functions for clarity

SageMoore commented on tests/utils.py:
```diff
@@ -567,9 +567,7 @@ def multi_process_parallel(
-    # NOTE: We need to set working_dir for distributed tests,
-    # otherwise we may get import errors on ray workers
-    ray.init(runtime_env={"working_dir": VLLM_PATH})
+    ray.init()
```
Nit: Can you clean this up? I assume you had some issue with setting the runtime_env but it does look like it's necessary.

SageMoore commented on vllm/distributed/device_communicators/custom_all_reduce.py:
```diff
@@ -175,16 +178,26 @@ def __init__(self,
-        lib = CudaRTLibrary()
-        pointer = lib.cudaMalloc(size_in_bytes)
-        handle = lib.cudaIpcGetMemHandle(pointer)
+        if current_platform.is_rocm():
```
All of this platform specific dispatching logic does make me think that we may be close to wanting to split CustomAllReduce into platform specific subclasses.

SageMoore commented on vllm/platforms/rocm.py:
```diff
@@ -5,6 +5,9 @@
 from typing import TYPE_CHECKING, Dict, List, Optional
 
 import torch
+from amdsmi import (AmdSmiException, amdsmi_get_gpu_asic_info,
```
Nit: looks like a few of these are unused. Can you trim down the import?

SageMoore reviewed (commented):
Thanks for the hard work on this, @ilmarkov. I made a first pass and left some comments.

Commit:
Fixes after review

Signed-off-by: ilmarkov <imarkov@redhat.com>

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -431,15 +507,20 @@ class CustomAllreduce {
   template <typename T>
   void allreduce(cudaStream_t stream, T* input, T* output, int size,
-                 int threads = 512, int block_limit = 36) {
+#if !defined(USE_ROCM)
```
Nit: Can you #define a BLOCK_LIMIT outside of this function declaration and do the USE_ROCM dispatching there?

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -362,7 +434,11 @@ class CustomAllreduce {
       // note: must share the base address of each allocation, or we get wrong
       // address
       if (cuPointerGetAttribute(&base_ptr,
+#if defined(USE_ROCM)
```
Same idea here. I generally think it's cleaner to do this kind of ifdef dispatching when declaring the constant variable and not in the implementation of a function.

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -237,7 +305,8 @@ DINLINE P* get_tmp_buf(Signal* sg) {
     cross_device_reduce_2stage(RankData* _dp, RankSignals sg, Signal* self_sg,
-                               T* __restrict__ result, int rank, int size) {
+                               T* __restrict__ result, int rank, int size,
+                               int compression_factor = 1) {
```
Nit: It looks like you aren't using the compression factor yet? If not, let's hold off adding it until it's used to eliminate any confusion in the short-term.

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -170,37 +178,97 @@ static DINLINE FlagType ld_flag_volatile(FlagType* flag_addr) {
+  if (threadIdx.x < ngpus) {
+    // simultaneously write to the corresponding flag of all ranks.
+    // Latency = 1 p2p write
+    __scoped_atomic_store_n(&sg.signals[threadIdx.x]->start[blockIdx.x][rank],
```
I assume that `rank` here is a remote rank? Meaning, the barrier has a round robin synchronization algorithm where each GPU is responsible for setting the count on exactly one other GPU? Then it waits for its own count to be set.

SageMoore reviewed (commented):
Left another round of comments. I generally feel good about this change for the AMD side of things but I'm a bit more hesitant to change the barrier logic on the Nvidia side. Assuming the algorithms are the same, I think we can run some Nvidia benchmarks and convince ourselves that we haven't regressed performance at all. CC @tlrmchlsmth

Commit:
ifdef and precommit fixes

Signed-off-by: ilmarkov <imarkov@redhat.com>

kahakuka commented:
@ilmarkov Hello, I am in version 5.7 of rocm. The 'hipIpcMemLazyEnabled PeerAccess' of the following function needs to be' 0 '. The accuracy is incorrect after changing to '0'. Excuse me, does this require a higher version of rocm to customize all reduce this time?

Commit:
Move MemIPC related primitives from python wrapper to cpp

Get rid of version dependency in rocm.
Remove rocm_wrapper because custom_allreduce does not need it anymore.

Signed-off-by: ilmarkov <imarkov@redhat.com>

kahakuka commented:
> @kahakuka感谢您的留言！您使用的是 MI200 GPU 吗？

Yes, I also tried it on MI250, it's the same.

ilmarkov commented on CMakeLists.txt:
```diff
@@ -520,6 +520,11 @@ if(VLLM_GPU_LANG STREQUAL "CUDA")
 # if CUDA endif
 endif()
 
+if(VLLM_GPU_LANG STREQUAL "HIP")
+  list(APPEND VLLM_EXT_SRC
+    "csrc/custom_all_reduce.cu")
+endif()
+
```
It goes right after the long CUDA `if` block. Do you think it makes sense to put HIP-related `if` block before CUDA?

tlrmchlsmth replied:
If you put it in this block, it will compile for both CUDA and HIP so no conditionals are needed

https://github.com/vllm-project/vllm/blob/2bb0e1a7994d55e964f5bb3acb79f8491fcb92f7/CMakeLists.txt#L229-L244

ilmarkov commented on tests/distributed/test_custom_all_reduce.py:
```diff
@@ -93,7 +93,7 @@ def eager_allreduce(tp_size, pp_size, rank, distributed_init_port):
     num_communication = rank // tp_size + 1
     sz = 1024
-    fa = get_tp_group().ca_comm
+    fa = get_tp_group().device_communicator.ca_comm
```
The `GroupCoordinator` (returned by `get_tp_group()`) does not have `ca_comm` field. Apparently, it was moved to `CudaCommunicator` class at some point and the test wasn't changed.

tlrmchlsmth commented on csrc/custom_all_reduce.cuh:
```diff
@@ -255,18 +336,20 @@ __global__ void __launch_bounds__(512, 1)
     tmps[i] = get_tmp_buf<P>(sg.signals[target]);
   }
   auto tmp_out = tmps[0];
-  multi_gpu_barrier<ngpus, true>(sg, self_sg, rank);
+  start_sync<ngpus>(sg, self_sg, rank);
```
Ok, I'll take another look in that case

ilmarkov commented on csrc/custom_all_reduce.cu:
```diff
@@ -142,3 +142,45 @@ void register_graph_buffers(fptr_t _fa,
   fa->register_graph_buffers(bytes, offsets);
 }
+
+std::tuple<fptr_t, torch::Tensor> allocate_shared_buffer_and_handle(
+    int64_t size) {
+  auto device_index = c10::cuda::current_device();
+  at::DeviceGuard device_guard(at::Device(at::DeviceType::CUDA, device_index));
+  void* buffer;
+  cudaStreamCaptureMode mode = cudaStreamCaptureModeRelaxed;
+  auto stream = c10::cuda::getCurrentCUDAStream().stream();
+  AT_CUDA_CHECK(cudaThreadExchangeStreamCaptureMode(&mode));
+#if defined(USE_ROCM)
+  // data buffers need to be "uncached" for signal on MI200
+  AT_CUDA_CHECK(
+      hipExtMallocWithFlags((void**)&buffer, size, hipDeviceMallocUncached));
+#else
+  AT_CUDA_CHECK(cudaMalloc((void**)&buffer, size));
+#endif
+
+  AT_CUDA_CHECK(cudaMemsetAsync(buffer, 0, size, stream));
+  AT_CUDA_CHECK(cudaStreamSynchronize(stream));
+  AT_CUDA_CHECK(cudaThreadExchangeStreamCaptureMode(&mode));
+
+  auto options =
+      torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCPU);
+  auto handle =
+      torch::empty({static_cast<int64_t>(sizeof(cudaIpcMemHandle_t))}, options);
+  AT_CUDA_CHECK(
+      cudaIpcGetMemHandle((cudaIpcMemHandle_t*)handle.data_ptr(), buffer));
+
+  return std::make_tuple(reinterpret_cast<fptr_t>(buffer), handle);
+}
```
Previously, the allocation and IPC initialization code was done in python wrapper of CUDA. However, in case of Rocm, there is a library version dependency on some of the constants (e.g. `hipIpcMemLazyEnabledPeerAccess`) fixing which in python would look ugly, in my opinion. So I decided to move the allocation code to cpp. 
I don't think it will affect the initialization performance.

I will prettify the function.

Commit:
Renaming and add comments

Signed-off-by: ilmarkov <imarkov@redhat.com>

kahakuka commented:
@ilmarkov Hello, may I ask another question? Is your test in eager mode or CUDAGraph mode. I tested the Eagle mode and found it to be normal, but there may be issues with the accuracy of the Cudagraph mode.

ilmarkov commented:
@kahakuka Hi, the `distributed/test_custom_all_reduce.py` verifies both eager and cudaGraph modes. Can you share the script so that we could reproduce the accuracy issue?

kahakuka commented:
> @kahakuka Hi, the `distributed/test_custom_all_reduce.py` test verifies both eager and cudaGraph modes. Can you share the script so that we could reproduce the accuracy issue?
@ilmarkov 
Thank you very much for your reply. I validated the accuracy through testing the model. The testing method is as follows:
Server side:
eager:  
HIP_VISIBLE_DEVICES=6,7 vllm serve /data/models/Qwen2-7B-Instruct  --enforce-eager --dtype float16 --trust-remote-code  -tp 2
cudagraph：
HIP_VISIBLE_DEVICES=6,7 vllm serve /data/models/Qwen2-7B-Instruct  --dtype float16 --trust-remote-code  -tp 2

client:
curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "/data/models/Qwen2-7B-Instruct",
        "messages": [
            {"role": "system", "content": "hello"}
        ]
    }'
The answer from eager is normal, while the response from CUDAGraph is garbled.

Commit:
Disable custom allreduce on older AMD GPUs

Signed-off-by: ilmarkov <imarkov@redhat.com>

ilmarkov commented:
@kahakuka I didn't manage to reproduce the issue on MI300X. So we decided to disable `custom_allreduce` on older AMD GPUs so that we could land this PR.

Commit:
Rename full_connected, fix test and
pre_commit error
Signed-off-by: ilmarkov <imarkov@redhat.com>

SageMoore commented on csrc/custom_all_reduce.cu:
```diff
@@ -12,7 +12,7 @@ static_assert(sizeof(void*) == sizeof(fptr_t));
 fptr_t init_custom_ar(const std::vector<fptr_t>& fake_ipc_ptrs,
                       torch::Tensor& rank_data, int64_t rank,
-                      bool full_nvlink) {
+                      bool full_connected) {
```
Nit: `fully_connected`

SageMoore commented on vllm/platforms/cuda.py:
```diff
@@ -101,7 +101,7 @@ def is_async_output_supported(cls, enforce_eager: Optional[bool]) -> bool:
 
     @classmethod
-    def is_full_nvlink(cls, device_ids: List[int]) -> bool:
+    def is_fully_connected_nvlink_or_xgmi(cls, device_ids: List[int]) -> bool:
```
NIT: `is_fully_connected` should be sufficient here.

SageMoore commented on vllm/distributed/device_communicators/custom_all_reduce.py:
```diff
@@ -72,6 +70,8 @@ def __init__(self,
         if not custom_ar:
             # disable because of missing custom allreduce library
             # e.g. in a non-cuda environment
+            logger.warning("Custom allreduce is disabled because "
```
This is a good addition.

tlrmchlsmth replied:
I think this should only be a warning on CUDA and RoCM platforms. IIUC this will always log a warning on other platforms and this should not be a warning e.g. on TPUs.

Also, could you change the comment on line 72 to `# e.g. in a non-GPU environment`

tlrmchlsmth replied:
Ok, I guess this is only constructed in the `cuda_communicator`, so not concerned about TPUs now.

However, it does look like this will always log a warning on AMD GPUs earlier than MI300X, and we shouldn't be warning users about expected behavior.

ilmarkov replied:
The custom allreduce library can be built on AMD GPU earlier than MI300X. We just don't enable it. So there shouldn't be a warning as we will not try to create `CustomAllreduce` object.

SageMoore commented on csrc/custom_all_reduce.cuh:
```diff
@@ -22,24 +27,38 @@
 };
 
 struct __align__(16) RankData {
-  const void* __restrict__ ptrs[8];
```
Can you leave a quick comment (in the PR not the code) as a reply to this comment explaining why `__restrict__` is no longer safe? Or was never safe in the first place?

SageMoore reviewed (approved):
Looks good to me, Ilia. All of my comments are just minor cosmetic things. I'll go ahead and approve.

kahakuka commented:
@ilmarkov 
Hello, the code interface functions for testing do not match the actual application. I'm not sure how you compiled them on your end.
<img width="330" alt="image" src="https://github.com/user-attachments/assets/2712845a-39c8-4b76-92dc-69a7cfa7666f" />

Commit:
Fix type check error

Signed-off-by: ilmarkov <imarkov@redhat.com>

ilmarkov commented:
@kahakuka custom allreduce requires P2P access between each pair of GPUs. For Nvidia, it is not required to have NVLink for that. On AMD xGMI guarantees  the P2P access so we only check that. It is possible to add a support for the cases without xGMI (but with p2p access) having checks similar to `gpu_p2p_access_check` in `custom_all_reduce_utils.py` but not as a part of this PR.

Commit:
Add comment

Signed-off-by: ilmarkov <imarkov@redhat.com>

russellb commented on vllm/config.py:
```diff
@@ -1537,11 +1537,13 @@ def _verify_args(self) -> None:
             from vllm.executor import ray_utils
             ray_utils.assert_ray_available()
-        if current_platform.is_rocm():
+        if (current_platform.is_rocm()
+                and current_platform.get_device_capability() is not None
+                and current_platform.get_device_capability() < (9, 4)):
```
This will fix CI for you:

```suggestion
        device_capability = current_platform.get_device_capability()
        if (current_platform.is_rocm() and device_capability is not None
                and device_capability < (9, 4)):
```

Commit:
Fix CI linter error

Signed-off-by: ilmarkov <imarkov@redhat.com>

tlrmchlsmth reviewed (approved):
Some small comments but LGTM otherwise -- I think we should downgrade that one `warning` log to `info`. WDYT?

tlrmchlsmth commented on csrc/custom_all_reduce.cuh:
```diff
@@ -22,24 +27,38 @@
+#ifndef USE_ROCM
+const int defaultBlockLimit = 36;
+CUpointer_attribute rangeStartAddrAttr =
+    CUDA_POINTER_ATTRIBUTE_RANGE_START_ADDR;
```
Looking at the build logs, I think this needs to be `CU_POINTER_ATTRIBUTE_RANGE_START_ADDR`

Commit:
Renaming and fixes

Signed-off-by: ilmarkov <imarkov@redhat.com>

gshtras commented on vllm/config.py:
```diff
@@ -1587,11 +1587,13 @@ def _verify_args(self) -> None:
-        if current_platform.is_rocm():
+        device_capability = current_platform.get_device_capability()
+        if (current_platform.is_rocm() and device_capability is not None
+                and device_capability < (9, 4)):
```
One small note here
device_capability should not be used on ROCm, as it is not sequential, as in bigger does not mean more capability, since on Radeon it is (11, x), or (12, x), but often times custom all reduce can not be used there

ilmarkov replied:
@gshtras Thank you for the note! Is there any way to enable it on MI300 and newer gpus other than checking `gcnArchName` of the device properties?

gshtras replied:
Afraid not, nothing that I'm aware of
The best way that I know of is something like https://github.com/vllm-project/vllm/blob/main/vllm/platforms/rocm.py#L265-L267
